### DISCUSSION TITLE
undefined method `edit_version' for Spaceship::ConnectAPI::App

### DISCUSSION THREAD

**Discussion by ChenChiaHung** [fastlane/fastlane#17024] [Q&A] [answered]
Hi all,

I looked at the [issue](https://github.com/fastlane/fastlane/issues/16644) and followed the [AppStoreConnect document](https://github.com/fastlane/fastlane/blob/master/spaceship/docs/AppStoreConnect.md) to implement a script of iOS submission, but the object returned by `Spaceship::ConnectAPI.get_app` does not support the method `edit_version`, this method is only supported by `Spaceship::Tunes::Application.find("xxx.xxx.xxx")` which will be deprecated. 

Thus, I'm trying to use `Spaceship::ConnectAPI::App` instead of `Spaceship::Tunes::Application.find("xxx.xxx.xxx")`, may I know whether there is a document about how to work with the similar method `edit_version` from `Spaceship::ConnectAPI` :-) 

Please correct me if I miss any information, thank you. :-)

fastlane 2.155.3

Spaceship::Tunes.login("username", "password")
app = Spaceship::ConnectAPI.get_app(app_id: 12345678).first
v = app.edit_version

**Comment by gsavit** [comment]
I'm wondering the same! There seem to be some changes to the new API and certain fields/functionality are missing?
Note that you can access some version info with a call to `Spaceship::ConnectAPI::App.find({appId})` and the `app_store_versions` field on the response, but those objects don't seem as complete as those returned via the old API.

**Reply by ChenChiaHung** [reply]
Hi @gsavit , you can have a look at the following information. :-)

**Comment by ChenChiaHung** [comment]
Hi @joshdholtz ,
Sorry for tagging you, can you please help to look into our question?
If anything we missed, please let us know.
Thank you :-)

**Comment by joshdholtz** [comment, marked as answer]
Here you go! https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/app.rb#L204

```rb
app = Spaceship::ConnectAPI::App.find(“com.bundle..id”)

# Spaceship::ConnectAPI::Platform::IOS
# Spaceship::ConnectAPI::Platform::MAC_OS
# Spaceship::ConnectAPI::Platform::TV_OS
# Spaceship::ConnectAPI::Platform::WATCH_OS
edit_version = app. get_edit_app_store_version(platform: Spaceship::ConnectAPI::Platform::IOS) 
```

**Reply by joshdholtz** [reply]
I’m still trying to find time to get documentation updated but there are always other things that come up 😔 `deliver` and `pilot` and completely migrated over to `Spaceship::ConnectAPI` so if you need examples on how to use stuff you _could_ like in there 🤷‍♂️ 

But feel free to keep pinging me in discussions and I’ll try to link to places for you to look at if you can’t find them!

**Reply by ChenChiaHung** [reply]
Thank you @joshdholtz , this is a great help to us, we will look into the information you provide. :-)

**Comment by ChenChiaHung** [comment]
Hi @joshdholtz, 
Sorry for bothering again, just a quick question for confirmation, a comment below mentions "not work yet" in the document "[AppStoreConnect document](https://github.com/fastlane/fastlane/blob/master/spaceship/docs/AppStoreConnect.md)", may I know is there any plan for supporting this function "create"? bcz I think this is a critical function to everyone for automating the iOS submission.

If I miss something, please let me know, thank you. :D

Create a new app
// Not working yet
app = Spaceship::ConnectAPI::App.create(name: "App Name",
                                        primary_language: "English",
                                        version_string: "1.0", # initial version
                                        sku: "123",
                                        bundle_id: "com.krausefx.app",
                                        platforms: ["IOS"])

**Reply by joshdholtz** [reply]
You can create an app in App Store Connect. It is being used in `produce` here 👉  https://github.com/fastlane/fastlane/blob/master/produce/lib/produce/itunes_connect.rb#L41-L49

I still need to add the ability to create the app in the Developer Portal (hoping I can get to this next week) but that will be done with something like `Spaceship::ConnctedAPI::BundleID.create(args args args)` 👈 might not actually be like this but that's the plan

**Reply by ChenChiaHung** [reply]
@joshdholtz , Thank you. :-)

**Comment by gsavit** [comment]
Thanks so much, Josh!

**Comment by ChenChiaHung** [comment]
Hi @joshdholtz 

Sorry for bothering you again, we have already done the part based on your useful information, it works well, but we still have the following partial code need to be revised, I'm not sure if there is any sample code or link to reference in the following parts? Thank you.

1. Set up the build
2. Add Translations
3. Set the submission information

```
# Set up the build
version.select_build(build) #We verified the build above, and now will be selecting the correct build.
version.release_on_approval = "false" #Manually release this version (for some weird reason, this needs to be in quotes, probably its is a string)
version.ratings_reset = false #Keep existing rating
version.phased_release["state"]["value"] = "INACTIVE" #Release update over 7-day period using phased release

#Add Translations
v.release_notes["en-US"] = enUS 
v.release_notes["en-GB"] = enUK
v.release_notes["en-AU"] = enUK
v.release_notes["ms"] = malay
v.release_notes["id"] = indonesian
v.release_notes["zh-Hant"] = chinese_traditional
v.release_notes["zh-Hans"] = chinese_simplified
v.save! #"Saved the app draft"

# Set app submission information
submission = app.create_submission
submission.add_id_info_uses_idfa=true #Does this app use the Advertising Identifier (IDFA)?
submission.add_id_info_serves_ads=true #Serve advertisements within the app
submission.add_id_info_tracks_install=true #Attribute this app installation to a previously served advertisement
submission.add_id_info_tracks_action=true #Attribute an action taken within this app to a previously served advertisement
submission.add_id_info_limits_tracking=true #Limit Ad Tracking setting in iOS
submission.complete! #This is the mother of all commands... (You are sending it for review)
```
